Prepend a question mark to the Azure Storage SAS token if it does not begin with one.

teo-tsirpanis opened the pull request:
[SC-35417](https://app.shortcut.com/tiledb-inc/story/35417/azure-bug-in-path-separator-and-query-string-prefix-usage-w-sas-tokens-with-azure-blob-storage)

Fixes TileDB-Inc/TileDB-Py#1844.

When the user specifies an Azure Storage SAS token, we pass it to the Azure SDK by appending it to the service endpoint URI. As shown in the Azure Portal, the SAS token starts with a `?`, but this is not always the case. This PR puts a `?` between the service endpoint and the SAS token if the SAS token does not start with a `?`.

In the linked GitHub issue, its reporter worked around it by prepending the SAS token with a `/?`, instead of `?`. Both [seem to be handled just fine in the Azure SDK](https://github.com/Azure/azure-sdk-for-cpp/blob/436c0905e20ec2892b5ff85fe9ada988ae24ebd7/sdk/core/azure-core/src/http/url.cpp#L73-L88).

Let me know how to test this.

---
TYPE: BUG
DESC: Support Azure Storage SAS tokens that don't start with a question mark.

davisp commented:
For testing, I'd just expose the `BlobServiceClient::GetUrl` [method](https://azuresdkdocs.blob.core.windows.net/$web/cpp/azure-storage-blobs/12.6.1/class_azure_1_1_storage_1_1_blobs_1_1_blob_service_client.html#a9cc5a9e05028ac608c17e50f4d379fc7). 

Once you've got that, then just assert that the returned value is `someting?${sas_token}` regardless of whether the URL or SAS has a trailing or leading `?` character.

teo-tsirpanis commented:
Test added. It sets a storage account name and either a SAS token with or without a `?` or no SAS token, and validates that `BlobServiceClient::GetUrl` returns the expected thing.

davisp reviewed (commented):
Do we want check to see if the user has configured their `vfs.azure.blob_endpoint` to include a `?`. Something like `if (blob_endpoint.find("?") != std::string::npos) throw std::invalid_argument("vfs.azure.blob_endpoint should not include a query string.");`

It seems bordering on silly to check given the setting name, but given how the docs are written I could see someone configuring their endpoint to be the whole url including the SAS in the query string.

teo-tsirpanis commented:
> configuring their endpoint to be the whole url including the SAS in the query string

That's how the new Azure SDK receives the SAS token. The `vfs.azure.storage_sas_token` option was initially removed in #3910 but reintroduced for compatibility in #4058.

IMHO we should warn if both the SAS token and the custom endpoint options are specified. The SAS token option is not really a thing with the new Azure SDK.

shaunrd0 reviewed (approved):
If the blob endpoint includes the SAS token and for some reason `vfs.azure.sas_token` is also set we could end up with `https://devstoreaccount1.blob.core.windows.net?baz=qux&foo=bar?baz=qux&foo=bar`, but azure decodes the SAS into a map so the repeated SAS didn't cause a problem when I tried. It ends up inserting the same `map[key]=value` twice, so GetUrl also still returns the correct `https://devstoreaccount1.blob.core.windows.net?baz=qux&foo=bar`.

https://github.com/Azure/azure-sdk-for-cpp/blob/436c0905e20ec2892b5ff85fe9ada988ae24ebd7/sdk/core/azure-core/src/http/url.cpp#L197

If the endpoint includes the SAS and `vfs.azure.sas_token` is not set we end up with  `https://devstoreaccount1.blob.core.windows.net?baz=qux&foo=bar?` which also doesn't cause an issue, but IMO we should do our best to get the blob endpoint correct for all combinations of these config settings even if Azure doesn't do something horrible for these two cases.

This would fix the issue though so :+1:

teo-tsirpanis commented:
Thanks for checking it out @shaunrd0! I missed the case when the SAS token is empty; we were still appending a question mark which the Azure SDK absorbs. I changed the code to cover that case, and added two more test cases.

davisp reviewed (approved):
+1
